### PR TITLE
Warn users when API token is used in Trusted Publishing project

### PR DESCRIPTION
**Pull request by facutuesca**
_(creating this draft PR here so that it can be reviewed with the `woodruffw-forks/ww/user-context` branch as base)_

If a user uploads files using an API token to a project that has a Trusted Publisher configured, send an email to all the project's users notifying them of the API token use, suggesting they remove it and use Trusted Publishing instead.

This warning is only sent once per (project, API token) combination.

This is a re-implementation of https://github.com/pypi/warehouse/pull/15453, taking into account the following requirements:
1. Send warnings for API tokens both project-scoped and account-level scoped.
2. Send only one email per `(API token, Project)` combination
3. Send the email to all of the project's maintainers
4. The email should be immediately actionable, including a link to the URL where the token owner can delete the token.

### CODE REVIEW

File: warehouse/packaging/models.py [hunk 1]
```diff
@@ -207,6 +208,19 @@ class Project(SitemapMixin, HasEvents, HasObservations, db.Model):
         passive_deletes=True,
     )
 
+    # Link to association table that keeps track if a warning was sent about use of an
+    # API token Macaroon in a Project with Trusted Publishing configured.
```
Comment (woodruffw) [inline warehouse/packaging/models.py]: Nitpicks:

```suggestion
    # Link to association table that keeps track of warnings for API 
    # token usage in projects that have Trusted Publishing configured.
```

File: warehouse/packaging/models.py [hunk 2]
```diff
@@ -874,3 +888,29 @@ class ProhibitedProjectName(db.Model):
     )
     prohibited_by: Mapped[User] = orm.relationship()
     comment: Mapped[str] = mapped_column(server_default="")
+
+
+class ProjectMacaroonWarningAssociation(db.Model):
```
Comment (woodruffw) [inline warehouse/packaging/models.py:893]: We can keep the `__tablename__`, but I don't think any of the other assoc tables in Warehouse include `Association` in the class name 🙂 

```suggestion
class ProjectMacaroonWarnings(db.Model):
```

Comment (facutuesca) [reply warehouse/packaging/models.py:893]: There's `OIDCProviderProjectAssociation` and `ObserverAssociation`

Comment (woodruffw) [reply warehouse/packaging/models.py:893]: Ah, whoops. Let's leave it then!

**Comment by facutuesca** [comment]
Closing this in favor of https://github.com/pypi/warehouse/pull/15641

**Pull request closed without merging**